runtime/metricsview/executor_timestamps.go

```diff
@@ -0,0 +1,199 @@
+package metricsview
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/rilldata/rill/runtime/drivers"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	defaultExecutionTimeout = time.Minute * 3
+)
+
+func (e *Executor) resolveDuckDBClickHouseAndPinot(ctx context.Context) (TimestampsResult, error) {
+	filter := e.security.RowFilter()
+	if filter != "" {
+		filter = fmt.Sprintf(" WHERE %s", filter)
+	}
+	timeDim := e.olap.Dialect().EscapeIdentifier(e.metricsView.TimeDimension)
+	escapedTableName := e.olap.Dialect().EscapeTable(e.metricsView.Database, e.metricsView.DatabaseSchema, e.metricsView.Table)
+
+	var watermarkExpr string
+	if e.metricsView.WatermarkExpression != "" {
+		watermarkExpr = e.metricsView.WatermarkExpression
+	} else {
+		watermarkExpr = fmt.Sprintf("MAX(%s)", timeDim)
+	}
+
+	rangeSQL := fmt.Sprintf(
+		"SELECT min(%[1]s) as \"min\", max(%[1]s) as \"max\", %[2]s as \"watermark\" FROM %[3]s %[4]s",
+		timeDim,
+		watermarkExpr,
+		escapedTableName,
+		filter,
+	)
+
+	rows, err := e.olap.Execute(ctx, &drivers.Statement{
+		Query:            rangeSQL,
+		Priority:         e.priority,
+		ExecutionTimeout: defaultExecutionTimeout,
+	})
+	if err != nil {
+		return TimestampsResult{}, err
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		var minTime, maxTime, watermark *time.Time
+		err = rows.Scan(&minTime, &maxTime, &watermark)
+		if err != nil {
+			return TimestampsResult{}, err
+		}
+		// Table is empty for the filters. duckdb returns null in this case
+		if minTime == nil {
+			return TimestampsResult{}, nil
+		}
+		return TimestampsResult{
+			Min:       *minTime,
+			Max:       *maxTime,
+			Watermark: *watermark,
```

Comment:
Dereferencing pointers without `nil` checks is not safe (maxTime and watermark are not checked). See comment above – I think the pointers can be avoided. Then the downstream logic can just check with `time.Time.IsZero()`